src/api/eth_subscribe.rs

```diff
@@ -0,0 +1,161 @@
+//! `Eth` namespace, subscriptions
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use api::Namespace;
+use futures::{Async, Future, Poll, Stream};
+use helpers::{self, CallResult};
+use rpc;
+use serde;
+use serde_json;
+use types::{Address, BlockHeader, Filter, H256, Log};
+use {DuplexTransport, Error};
+
+/// `Eth` namespace, subscriptions
+#[derive(Debug, Clone)]
+pub struct EthSubscribe<T> {
+  transport: T,
+}
+
+impl<T: DuplexTransport> Namespace<T> for EthSubscribe<T> {
+  fn new(transport: T) -> Self
+  where
+    Self: Sized,
+  {
+    EthSubscribe { transport }
+  }
+
+  fn transport(&self) -> &T {
+    &self.transport
+  }
+}
+
+pub struct SubscriptionStream<T: DuplexTransport, I> {
+  transport: T,
+  id: String,
+  rx: T::NotificationStream,
+  _marker: PhantomData<I>,
+}
+
+impl<T: DuplexTransport, I> SubscriptionStream<T, I> {
+  pub fn new(transport: T, id: String) -> Self {
+    let rx = transport.subscribe(&id);
+    SubscriptionStream {
+      transport,
+      id,
+      rx,
+      _marker: PhantomData,
+    }
+  }
+
+  pub fn id(&self) -> &str {
+    &self.id
+  }
+}
+
+impl<T, I> Stream for SubscriptionStream<T, I>
+where
+  T: DuplexTransport,
+  I: serde::de::DeserializeOwned,
+{
+  type Item = I;
+  type Error = Error;
+
+  fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+    match self.rx.poll() {
+      Ok(Async::Ready(Some(x))) => serde_json::from_value(x)
+        .map(Async::Ready)
+        .map_err(Into::into),
+      Ok(Async::Ready(None)) => Ok(Async::Ready(None)),
+      Ok(Async::NotReady) => Ok(Async::NotReady),
+      Err(e) => Err(e),
+    }
+  }
+}
+
+impl<T: DuplexTransport, I> fmt::Debug for SubscriptionStream<T, I> {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "{:?}", self.id())
+  }
+}
+
+impl<T: DuplexTransport, I> Drop for SubscriptionStream<T, I> {
+  fn drop(&mut self) {
+    self.transport.unsubscribe(self.id());
```

Comment:
In ticket for #37 , desired behavior was to unsubscribe from notifications on dropping the stream. This "unregisters" this stream from the transport, but does not invoke `eth_unsubscribe` via rpc. Unsure how to best handle this, thought waiting on a future from drop would be a bad idea. Currently if a user forgets to call eth_unsubscribe it's not fatal but notifications will continue to be sent via the transport and not forwarded to a stream.

Reply:
Can we get a `pub fn unsubscribe(self)` on the stream to allow people to unsubscribe?